scrapy/utils/iterators.py

```diff
@@ -71,30 +101,39 @@ def xmliter_lxml(obj, nodename, namespace=None, prefix="x"):
 
 
 class _StreamReader:
-    def __init__(self, obj):
-        self._ptr = 0
-        if isinstance(obj, Response):
+    def __init__(self, obj: Union[TextResponse, str, bytes]):
+        self._ptr: int = 0
+        self._text: Union[str, bytes]
+        if isinstance(obj, TextResponse):
             self._text, self.encoding = obj.body, obj.encoding
         else:
             self._text, self.encoding = obj, "utf-8"
-        self._is_unicode = isinstance(self._text, str)
+        self._is_unicode: bool = isinstance(self._text, str)
 
-    def read(self, n=65535):
-        self.read = self._read_unicode if self._is_unicode else self._read_string
+    def read(self, n: int = 65535) -> bytes:
+        self.read: Callable[[int], bytes] = (  # type: ignore[method-assign]
```

Comment:
This code is wild: if you look closely you'll see that it assigns a different callable to itself. I have no idea if it's intended. I even think the behavior on the first and subsequent calls is different as `lstrip()` is only called on the first call.

Reply:
Maybe we should take the chance to fix it? e.g. assign `self._read = …` on `__init__`, and use `self._read(n).lstrip()` here?

Reply:
I think it could just use a local variable and not assign it anywhere. But, as usual, I wonder if fixing it will break anytning :)

On the other hand, it shouldn't and we have tests for it.

Reply:
Looks like it intentionally only strips on the first read, to skip the leading whitespace in the file (it's even possible that it's just for tests which have leading whitespace). `xmliter` instead uses regexes to find the document header.

In any case, we don't even advertise this function, even though it's used in some old 3rd-party code, I don't want to spend too much time on it.